### PR TITLE
Rename REST API endpoints to make the interface more RESTful.

### PR DESCRIPTION
**Pull request by zinigor**
### Changes proposed in this Pull Request:
Moving the endpoints so that the interface looks like this:
### The module group:
- [x] GET `/module` returns all available modules and their status;
- [x] GET `/module/(?P[a-z\-]+)` returns a single module information;
- [x] GET `/module/(?P[a-z\-]+)/active` returns module’s activation status;
- [x] POST `/module/(?P[a-z\-]+)/active` sets module’s activation status;
- [x] POST `/module/(?P[a-z\-]+)/` updates module options;
- [x] POST `/module/activate` activates a bunch of modules – perhaps we should use a slug like all here?
- [x] GET `/modules/(?P[a-z\-]+)/data` retrieves special data that a module provides, like VaultPress data, Monitor stats, Protect count, verification tools keys, etc;

### The connection group:
- [x] GET `/connection` gives you connection status;
- [x] GET `/connection/url` gives you an URL you need to go to connect Jetpack;
- [x] GET `/connection/data` gives you connection data;
- [x] DELETE `/connection` disconnects Jetpack
- ~~GET `/connection/ssl` re-checks SSL status;~~
- [x] DELETE `/connection/user` disconnects a user from dotcom;

### The miscellaneous group:
- [x] DELETE `/options/` resets an option or options that are passed as an array;
- [x] POST `/jumpstart` activates or deactivates jumpstart depending on the passed boolean;
- [x] DELETE `/notice/` dismisses a certain notice;
- [x] GET `/settings/` retrieves miscellaneous settings;
- [x] POST `/settings` sets miscellaneous settings.

### CODE REVIEW

**Comment by eliorivero** [comment]
Changed the
`DELETE /connection/mine`
to
`DELETE /connection/user`
since the endpoint is generic and can be used to unlink any user, not just my user.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 1]
```diff
@@ -14,20 +14,934 @@ class Jetpack_Core_API_Module_Activate_Endpoint
...
 	}
+}
+
+class Jetpack_Core_API_Module_List_Endpoint {
+
+	public function process( $request ) {
+		if ( 'GET' === $request->get_method() ) {
+			return $this->get_modules( $request );
+		} else {
+			return $this->update_modules( $request );
```
Comment (eliorivero) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: This method doesn't exist.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 2]
```diff
@@ -14,20 +14,934 @@ class Jetpack_Core_API_Module_Activate_Endpoint
...
+				case 'monitor_receive_notifications':
+					$monitor = new Jetpack_Monitor();
+
+					// If we got true as response, consider it done.
+					$updated = true === $monitor->update_option_receive_jetpack_monitor_notification( $value );
+					break;
+
+				case 'post_by_email_address':
+					if ( 'create' == $value ) {
+						$result = $this->_process_post_by_email(
```
Comment (eliorivero) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Trying to perform regeneration on Post By Email throws:
```php
<br />
<b>Fatal error</b>:  Using $this when not in object context in <b>/public_html/wp-content/plugins/jetpack/_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php</b> on line <b>384</b><br />
```

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 3]
```diff
@@ -14,20 +14,934 @@ class Jetpack_Core_API_Module_Activate_Endpoint
...
+	 * @return array
+	 */
+	private function _process_post_by_email( $endpoint, $error ) {
+		if ( ! current_user_can( 'edit_posts' ) ) {
+			return array( 'message' => $error );
+		}
+
+		$this->xmlrpc->query( $endpoint );
+
+		if ( $this->xmlrpc->isError() ) {
```
Comment (eliorivero) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: This `xmlrpc` property is private in the parent class so it's inaccessible from the children.

**Comment by eliorivero** [comment]
I'm leaving the 
> GET /connection/ssl re-checks SSL status;
as it is, since we don't have an endpoint for that. You told me that this was for
> the endpoint that looks something like `ssl/force-recheck`
but doing a search for `register_rest_route` yields no results.
There is something related: the `wp_ajax_jetpack-recheck-ssl` hook, but since we don't have an endpoint for that, we should dismiss it for now.

**Comment by oskosk** [comment]
@zinigor: I renamed your  proposed `POST /jumpstart/active` to just `POST /jumpstart`. 

It was by mistake at first, but then I thought that having an `active` both in the URL and in the body of the POST request, could be redundant.

**Comment by eliorivero** [comment]
@zinigor I've rebased this branch to solve a conflict in the file `_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php` so you might want to review it.
As a backup, I saved the previous state before the rebasing in this branch https://github.com/Automattic/jetpack/tree/rename%2Fapi-endpoints-save

**Comment by oskosk** [comment]
@eliorivero @zinigor I needed to force push after @eliorivero's rebase as I had implemented the `DELETE /options` incorrectly in my first commit. nothing bad happened, just letting you know.

**Comment by zinigor** [comment]
Thank you wonderful people! I have added some final changes to make it all work, now we seem to be ready to go.

**Comment by eliorivero** [comment]
Added
```scss
"atoms/typography/functions",
"atoms/typography/variables",
```
to `/scss/jetpack-admin.scss` to prevent errors:
```shell
Error in plugin 'sass'
Message:
    scss/templates/_main.scss
Error: Undefined variable: "$gill".
        on line 446 of scss/templates/_main.scss
>> 	font-family: $gill;
   --------------^
```
```shell
Error in plugin 'sass'
Message:
    scss/templates/_connection-landing.scss
Error: invalid operands for multiplication
        on line 8 of scss/templates/_connection-landing.scss
>> $j-gutter: em(20px);
   -----------^
```

**Comment by oskosk** [comment]
Let's merge this when builds finish :shipit: 🚢 LGTM

**Comment by oskosk** [comment]
@zinigor did this one get implemented ?

* GET /module returns all available modules and their status;

**Pull request merged**